staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -693,19 +698,62 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
 	return false
 }
 
-func ensureNodeProvidedIPExists(node *v1.Node, nodeAddresses []v1.NodeAddress) (*v1.NodeAddress, bool) {
-	var nodeIP *v1.NodeAddress
-	nodeIPExists := false
-	if providedIP, ok := node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr]; ok {
-		nodeIPExists = true
-		for i := range nodeAddresses {
-			if nodeAddresses[i].Address == providedIP {
-				nodeIP = &nodeAddresses[i]
-				break
+// preferProvidedNodeIP re-orders node addresses, placing a node IP suggested by
+// the user first among node addresses of the same type.
+// If the user suggested a node IP which is not present in node addresses it
+// will return an error.
+func preferProvidedNodeIP(node *v1.Node, nodeAddresses []v1.NodeAddress) ([]v1.NodeAddress, error) {
+	providedIP, ok := node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr]
+	if !ok {
+		// No suggested node IP
+		return nodeAddresses, nil
+	}
+	klog.V(2).Infof("Kubelet provided node IP %s. Ensuring it is listed as first node address.", providedIP)
+
+	findNodeIPIndex := func(address string) (int, bool) {
+		for i, nodeAddress := range nodeAddresses {
+			if nodeAddress.Address == address {
+				return i, true
 			}
 		}
+		return 0, false
+	}
+
+	nodeIPIndex, nodeIPFound := findNodeIPIndex(providedIP)
```

Comment:
this is only reordering a single record... the cloud provider could return multiple addresses with Address=<nodeIP> with different types

Reply:
Oh, that's fun.

Reply:
Thinking about it, I think we should confine this behaviour to internal and external IP. I don't think it's relevant to the others? However, I can implement re-ordering for both if they both match.

Reply:
I don't really have context to know what the desired/expected behavior is, but scoping re-ordering to a particular type/types could be reasonable...

Reply:
This can now re-order multiple type matches, but explicitly confines the behaviour to internal and external IP.